torch/_dynamo/variables/misc.py

```diff
@@ -234,11 +234,44 @@ def visit(node):
 
         VariableTracker.apply(visit, (args, kwargs))
 
+        ctx = AutogradFunctionContextVariable.create(tx)
+        args = [ctx, *args]
+
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            from .torch import is_fn_safe_to_run, TorchHigherOrderOperator
+
+            def trampoline_autograd_apply(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            def trampoline_autograd_fwd(*args, **kwargs):
+                return self.fn_cls.forward(*args, **kwargs)
+
+            def trampoline_autograd_bwd(*args, **kwargs):
+                return self.fn_cls.backward(*args, **kwargs)
+
+            # Speculate fwd, will raise unimplemented and bubble up if not sound, or will accumulate guards
+            # onto tx if sound.
+            # TODO(voz): NOTE: This function identity is unguarded, but the odds of someone swapping self.fn_cls from
+            # autograd fn to something else is very low.
+            higher_order_autograd_fn = TorchHigherOrderOperator(trampoline_autograd_fwd)
+            speculated_fwd_result = higher_order_autograd_fn.call_function(
+                tx, args, kwargs
+            )
+            bwd_args = [ctx, speculated_fwd_result]
+            # ctx.value.saved_tensors = ctx.value.to_save
+            if not is_fn_safe_to_run(
+                tx, TorchHigherOrderOperator(trampoline_autograd_bwd), bwd_args
+            ):
```

Comment:
Why is this not just `is_fn_safe_to_run(tx, trampoline_autograd_bwd, bwd_args)`?

Reply:
We need to make it an operator so we get the higher order operator behavior on call_function, the one that does the actual speculation. 

We could of course, hide that in is_fn_safe_to_run.

Reply:
That makes sense. No need to hide unless you think that would be nicer.